src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java

```diff
@@ -535,20 +537,26 @@ public List<Path> exportStudentRepositories(ProgrammingExercise programmingExerc
                     programmingExercise.getTitle(), participations.stream().map(StudentParticipation::getParticipantIdentifier).collect(Collectors.joining(", ")));
         }
 
-        List<Path> exportedStudentRepositories = new ArrayList<>();
-        participations.forEach(participation -> {
+        List<Path> exportedStudentRepositories = Collections.synchronizedList(new ArrayList<>());
+
+        log.info("export student repositories for programming exercise {} in parallel", programmingExercise.getId());
+        var threadPool = Executors.newFixedThreadPool(10);
+        var futures = participations.stream().map(participation -> CompletableFuture.runAsync(() -> {
             try {
+                log.debug("invoke createZipForRepositoryWithParticipation for participation {}", participation.getId());
                 Path zipFile = createZipForRepositoryWithParticipation(programmingExercise, participation, repositoryExportOptions, workingDir, outputDir);
                 if (zipFile != null) {
                     exportedStudentRepositories.add(zipFile);
                 }
             }
-            catch (Exception e) {
+            catch (Exception exception) {
                 var error = "Failed to export the student repository with participation: " + participation.getId() + " for programming exercise '" + programmingExercise.getTitle()
                         + "' (id: " + programmingExercise.getId() + ") because the repository couldn't be downloaded. ";
                 exportErrors.add(error);
             }
-        });
+        }, threadPool).toCompletableFuture()).toList().toArray(new CompletableFuture<?>[participations.size()]);
```

Comment:
We had a PR earlier that replaced such calls with a zero with array claiming to improve performance and synchronization issues.
See #5885

Reply:
```suggestion
        }, threadPool).toCompletableFuture()).toArray(CompletableFuture[]::new);
```